### PR TITLE
diagnostics: Support oc adm diagnostics securitycontextconstraints

### PR DESCRIPTION
**Pull request by nak3**
This patch adds oc adm diagnotics scc to check that the default
securitycontextconstraints are present and contain the expected
permissions.

The basic concept is same with `diagnostics clusterrolebindings` and
`diagnostics clusterrolebindings`. Having said that , scc have many
diffable rules, so clusterscc collects diffs both `Union = true` and
`Union = false` and informs as Info and Debug level respectively.

### CODE REVIEW

**Comment by simo5** [comment]
/assign @php-coder 
Slava PTAL

**Comment by nak3** [comment]
@simo5 Thank you. Updated message issues.

As for the test, I am thinking that the test codes should use utils in [pkg/oc/admin/diagnostics/diagnostics/cluster/aggregated_logging/diagnostic_test.go](https://github.com/openshift/origin/blob/9dae2335a826caeeb19c60503df556617a2817a2/pkg/oc/admin/diagnostics/diagnostics/cluster/aggregated_logging/diagnostic_test.go), but these are private. Or these should be moved to util packages.
I opened https://github.com/openshift/origin/issues/19729 instead of fixing here.

File: pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	securitytypedclient "github.com/openshift/origin/pkg/security/generated/internalclientset/typed/security/internalversion"
+)
+
+// ClusterSCC is a Diagnostic to check that the default securitycontextconstraints match expectations
+type ClusterSCC struct {
+	SCCClient securitytypedclient.SecurityContextConstraintsInterface
+	SARClient authorizationtypedclient.SelfSubjectAccessReviewsGetter
+}
+
+const SccName = "ClusterSCC"
```
Comment (enj) [inline pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: Why not SCC?

Comment (nak3) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: I standarized the name under `pkg/oc/admin/diagnostics/diagnostics/cluster` with `ClusterXXX`, after reviewed other diagnostics.

(e.g)
https://github.com/openshift/origin/blob/ea90a090da768e9a505b0ca2368927d7d25210f1/pkg/oc/admin/diagnostics/diagnostics/cluster/registry.go#L31

https://github.com/openshift/origin/blob/ea90a090da768e9a505b0ca2368927d7d25210f1/pkg/oc/admin/diagnostics/diagnostics/cluster/router.go#L32


But it looks like `ClusterXXX` was not necessary(?) I realized that some did not have it.

https://github.com/openshift/origin/blob/ea90a090da768e9a505b0ca2368927d7d25210f1/pkg/oc/admin/diagnostics/diagnostics/cluster/service_externalip.go#L28

Comment (nak3) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: I think that `ClusterSCC` shoud be used here. In case someone wants to add diagnostics related to `scc` in the future, (if I used `SCC` here), `SCC` will include wide meaning and may cause confusion.

Comment (enj) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: Then the correct approach would be to have a top level scc check with this as a child.  ClusterSCC makes no sense because SCC has not namespace equivalent like RBAC's Cluster/Role.

Comment (nak3) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: > ClusterSCC makes no sense because SCC has not namespace equivalent like RBAC's Cluster/Role.

Since SCC is effected by SA, SCC has a relation to namespace and sometimes need namespace based check. Actually, logging namespace has its SCC diagnostics as: https://github.com/openshift/origin/blob/8ac6446f8e4f2db4dea4448b36de7af69ee6860b/pkg/oc/admin/diagnostics/diagnostics/cluster/aggregated_logging/scc.go

Comment (enj) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: Sure, but the SCC check itself is always cluster scoped.  It does not matter that the user bound to it exists only in a namespace.

Comment (php-coder) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: +1 for renaming it to be "SCC".

File: pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+	"github.com/openshift/origin/pkg/cmd/server/bootstrappolicy"
+	"github.com/openshift/origin/pkg/oc/admin/diagnostics/diagnostics/types"
+	"github.com/openshift/origin/pkg/oc/admin/diagnostics/diagnostics/util"
+	policycmd "github.com/openshift/origin/pkg/oc/admin/policy"
+	securityapi "github.com/openshift/origin/pkg/security/apis/security"
+	securitytypedclient "github.com/openshift/origin/pkg/security/generated/internalclientset/typed/security/internalversion"
+)
+
+// ClusterSCC is a Diagnostic to check that the default securitycontextconstraints match expectations
```
Comment (php-coder) [inline pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: Consider improving this comment to briefly describe these expectations. Otherwise this comment looks useless and produces more questions rather than making something clear.

P.S. godoc comment should also end with dot AFAIR.

File: pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+func (d *ClusterSCC) Description() string {
+	return "Check that the default SecurityContextConstraints are present and contain the expected permissions"
+}
+
+func (d *ClusterSCC) Requirements() (client bool, host bool) {
+	return true, false
+}
+
+func (d *ClusterSCC) CanRun() (bool, error) {
+	if d.SCCClient == nil {
```
Comment (php-coder) [inline pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: Did you want to check `d.SARClient` here?

File: pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	r := types.NewDiagnosticResult(SccName)
+	reconcileOptions := &policycmd.ReconcileSCCOptions{
+		Confirmed:      false,
+		Union:          true,
+		Out:            ioutil.Discard,
+		SCCClient:      d.SCCClient,
+		InfraNamespace: bootstrappolicy.DefaultOpenShiftInfraNamespace,
+	}
+
+	changedSCCs, err := reconcileOptions.ChangedSCCs()
```
Comment (php-coder) [inline pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: Do we need to invoke `ReconcileSCCOptions.Validate()` to ensure that everything has been configured correctly?

File: pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	if err != nil {
+		r.Error("CSD1000", err, fmt.Sprintf("Error inspecting SCCs: %v", err))
+		return r
+	}
+	changedSCCNames := map[string]bool{}
+	for _, changedSCC := range changedSCCs {
+		_, err := d.SCCClient.Get(changedSCC.Name, metav1.GetOptions{})
```
Comment (php-coder) [inline pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: It looks like `ChangedSCCs()` could be more useful if it would return 2 different sets of SCCs -- one for non-existing and another one that need to be updated. Perhaps, it can be implemented as a separate PR..

File: pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go (hunk 2 above)
Comment (nak3) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: I used this comment based on ClusterRole's diagnostics:

https://github.com/openshift/origin/blob/12550a638027fb4b71332162f1d459a9d7d3c1c7/pkg/oc/admin/diagnostics/diagnostics/cluster/roles.go#L21

I will update `roles.go` if updated comment is fine.

File: pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go (hunk 4 above)
Comment (nak3) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: I think it is not necessary. `ReconcileSCCOptions.Validate()` https://github.com/openshift/origin/blob/12550a638027fb4b71332162f1d459a9d7d3c1c7/pkg/oc/admin/policy/reconcile_sccs.go#L140 is designed for the validation of command `oc adm policy reconcile-sccs`. If this diagnostics scc also called, it may cause some error in the future when `Validate()` was updated only for `oc adm policy reconcile-sccs`'s validation.

File: pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go (hunk 5 above)
Comment (nak3) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: Yes, that's useful. But it would be great if it is implemented as a separate PR.

File: pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go (hunk 1 above)
Comment (nak3) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: I understand. Updated to "SCC"

**Comment by nak3** [comment]
Thank you @php-coder @enj Updated.

File: pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go (hunk 4 above)
Comment (php-coder) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: On the other hand, if they add some member that will be used by `ChangedSCCs()`, we will notice that fact very late.

Comment (nak3) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: I understand and agree with you. But `Validate()` in `reconcile-sccs.go` uses `NSClient`, which needs `clientcmd.Factory`. I am wondering that bringing `clientcmd.Factory` into here is not appropriate.

https://github.com/openshift/origin/blob/12550a638027fb4b71332162f1d459a9d7d3c1c7/pkg/oc/admin/policy/reconcile_sccs.go#L144-L146

**Comment by php-coder** [comment]
@nak3 Another question: why we need this and why/where it's useful? We have `oc adm policy reconcile-sccs` command that already do the same thing by default, so how this command is different?

**Comment by nak3** [comment]
Actually, not only `oc adm policy reconcile-sccs`, but also all of `oc adm diagnotics` subcommands have alternative commands 😭 So, I will explain that the advantage of `oc adm diagnotics`.

In my support experience, some customers run `oc adm diagnostics` regular basis or after they performed some operations like install, upgrade, or make some changes. So, `oc adm diagnostics` is the tool that they can find some issues in advance.

Also, `oc adm diagnotics` can cover all of basic check for troubleshootings by only one command, instead of using list of alternative commands like `oc adm policy reconcile-sccs`, `oc adm policy reconcile-cluster-role`, `oc adm policy reconcile-cluster-role-bindings`, `oc status` etc ... It is sometimes useful for the troubleshooting by openshift newbies.

**Comment by php-coder** [comment]
@nak3 If it's possible, could you squash the commits? Into single one, or even better into 2 (one with your changes and another one with autogenerated changes).

**Comment by nak3** [comment]
Sure, squashed. I squashed only 1, but if I should separate it into two, please let me know.

**Comment by php-coder** [comment]
@nak3 It's ok, thank you!

Looks like you need also to run `hack/update-generated-docs.sh` at least to fix the build.

**Comment by php-coder** [comment]
I'm not sure whether it's in the scope of this PR or it isn't but: I think that it we also need to warn about SCC that has been granted to all authenticated users. It's not fatal error of course but
1) it shows that cluster admin doesn't care/know about security much
2) it could simplify diagnostic of some strange cases where less strict SCCs (like `anyuid`) affects system pods.

I'd suggest to implement this check in the following way:
1) for each standard SCCs except "restricted": warn if it's granted to `system:authenticated` group
2) for any SCCs (including custom ones): warn if it's granted to `system:unauthenticated` group or `system:anonymous` user

WDYT?

**Comment by php-coder** [comment]
> It would be great if you could allow me to implement it by another PR.

Yes, sure.

Thank you for example with `nonroot` SCC!

**Comment by nak3** [comment]
Sure, separated commit.

File: pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	securitytypedclient "github.com/openshift/origin/pkg/security/generated/internalclientset/typed/security/internalversion"
+)
+
+// SCC is a Diagnostic to check that the default securitycontextconstraints are present and contain the expected permissions.
+type SCC struct {
+	SCCClient securitytypedclient.SecurityContextConstraintsInterface
+	SARClient authorizationtypedclient.SelfSubjectAccessReviewsGetter
+}
+
+const SccName = "SCC"
```
Comment (soltysh) [inline pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: Shouldn't we use full resource name and additionally allow using shorter aliases? That's the approach we're taking across the entire CLI, I don't think this should be any different.

Comment (nak3) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: I see. I basically agree that full resource name with shorter alias.

However, subcommands of diagnostics set their alias in one same code here:

https://github.com/openshift/origin/blob/e9277a4c9f09115a13a719c78afef0ef90736ed8/pkg/oc/admin/diagnostics/diagnostics.go#L179

So, if I try to let it use `securitycontextcontraints` with `scc` alias, I need to change above code. An all of subcommands of diagnostics need to be updated a little bit.

Comment (enj) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: I will defer these CLI specifics to @soltysh as I have no strong opinions.

Comment (soltysh) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: @nak3 then just do full name (like with `oc adm diagnostics clusterrolebindings`) and don't mention short names at all.

Comment (nak3) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: @soltysh Sure, updated.

**Comment by nak3** [comment]
@soltysh It would be great if you could take a look.

**Comment by soltysh** [comment]
/hold cancel
@nak3 your changes lgtm, I'll merge it but please squash your changes into two commits (actual changes + generated)

**Comment by enj** [comment]
/hold

@soltysh today is code freeze for 3.10 right?

@nak3 please update the title of the PR and the first message to reflect the final state.  Include this updated information in the commit message as well.

**Comment by nak3** [comment]
@soltysh @enj Thank you. Updated.

**Comment by soltysh** [comment]
> @soltysh today is code freeze for 3.10 right?

It is now ;-) I was hoping to merge it for 3.10, but it'll have to wait for the next release in that case.

**Comment by php-coder** [comment]
> it'll have to wait for the next release in that case

:-( It's a dedicated command, how it may hurt?

**Comment by enj** [comment]
The code freeze guidelines are extremely clear.  This is not a critical fix so it needs to wait.

File: pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go (hunk 5 above)
Comment (php-coder) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: @stlaz Would you be interested in implementing such a refactoring?

Comment (stlaz) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: Sure!

**Comment by php-coder** [comment]
It still
/lgtm

**Comment by enj** [comment]
@php-coder @stlaz please remove the hold if you are happy with this.

**Comment by php-coder** [comment]
@nak3 Thank you again for the contribution and I'm sorry that it took so long time to get merged.

/hold cancel
/lgtm

**Comment by php-coder** [comment]
@nak3 AFAIU the command was renamed from `oc adm diagnotics scc` to `oc adm diagnotics securitycontextconstraints` along the way? If yes, could I ask you to edit the commit messages and PR description to reflect this? Thanks!

**Comment by enj** [comment]
I just spoke with @deads2k in regards to the rebase.  No changes can be made to *any* `oc` code until after it lands.  Sorry @nak3, more waiting ahead :/

**Comment by nak3** [comment]
@php-coder @enj No problem. Updated.

File: pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go (hunk 5 above)
Comment (stlaz) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: I've had the refactoring ready for quite some time now. Given that we're waiting for the rebase, should I just post it to this PR or transitively wait for the rebase to happen so that this PR is merged and create a new PR with it afterwards?

Comment (php-coder) [reply pkg/oc/admin/diagnostics/diagnostics/cluster/scc.go]: I'd rather base your PR on top of this one.

**Comment by deads2k** [comment]
rebase has landed.

**Comment by stlaz** [comment]
@nak3 could you please rebase your PR (hopefully for the last time)?

**Comment by stlaz** [comment]
@nak3 when I tried to rebase my own PR, the changes here did not seem to be picked up by `./hack/update-generated-completions.sh`. Can you please rebase and check that everything works as it should?

**Comment by nak3** [comment]
@stlaz I'm sorry! I missed your notification. Rebased now.
